README.md
# NodeJS [SwiftPay](https://swiftpayru.docs.apiary.io/#) API

[![license](https://img.shields.io/npm/l/swiftpay-api?style=flat-square)](https://github.com/xTCry/node-swiftpay-api/blob/main/LICENSE)
[![npm](https://img.shields.io/npm/v/swiftpay-api?style=flat-square)](https://npmjs.com/package/swiftpay-api)
[![downloads](https://img.shields.io/npm/dm/swiftpay-api?style=flat-square)](https://npmjs.com/package/swiftpay-api)
[![GitHub](https://img.shields.io/github/stars/xTCry/node-swiftpay-api?style=flat-square)](https://github.com/xTCry/node-swiftpay-api)
[![last commit](https://img.shields.io/github/last-commit/xTCry/node-swiftpay-api?style=flat-square)](https://github.com/xTCry/node-swiftpay-api)

## 📦 Установка

```shell
yarn add swiftpay-api
```

## 🛠️ Использование

```typescript
import 'dotenv/config';
import { ApiClient } from 'swiftpay-api';

const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));

const main = async () => {
  const api = new ApiClient(
    process.env.WALLET_PROVIDER_SWIFTPAY_API_KEY!,
    Number(process.env.WALLET_PROVIDER_SWIFTPAY_SHOP_ID!),
  );

  let time = Date.now();
  const stats = await api.getStats();
  console.log('stats', stats);
  console.log(`Time: ${Date.now() - time} ms`);

  await sleep(500);

  time = Date.now();
  const account = await api.getAccount();
  console.log('account', account);
  console.log(`Time: ${Date.now() - time} ms`);

  await sleep(500);

  time = Date.now();
  const shops = await api.getShops();
  console.log('shops', shops);
  console.log(`Time: ${Date.now() - time} ms`);

  await sleep(500);

  if (process.env.WALLET_PROVIDER_SWIFTPAY_SHOP_ID) {
    time = Date.now();
    const { data: shop } = await api.getShop();
    console.log('shop', shop);
    console.log(`Time: ${Date.now() - time} ms`);
  }

  await sleep(500);

  // time = Date.now();
  // const { data: createOrder } = await api.createOrder({
  //   amount: 100,
  //   order_id: 4e7 + 2,
  //   data: { userId: 1230, email: 'gmail' },
  // });
  // console.log('createOrder', createOrder);
  // console.log(`Time: ${Date.now() - time} ms`);

  // await sleep(500);

  time = Date.now();
  const orders = await api.orders({ sort: 'id', data: [] });
  console.log('orders', orders);
  console.log(`Time: ${Date.now() - time} ms`);
};

main()
  .then()
  .catch((e) => {
    console.error(e);
  });


```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.0.5](https://github.com/xTCry/node-swiftpay-api/compare/v1.0.4...v1.0.5) (2022-02-01)


### Bug Fixes

* **api:** fixed undefined `response` ([fe2c566](https://github.com/xTCry/node-swiftpay-api/commit/fe2c566c88a7672f38b29795a28467f5d9e35ddf))

### [1.0.4](https://github.com/xTCry/node-swiftpay-api/compare/v1.0.3...v1.0.4) (2022-01-26)


### Bug Fixes

* **types:** renamed `PaymentResponse` ([04f2e2d](https://github.com/xTCry/node-swiftpay-api/commit/04f2e2d57da386def340591fb0e6c233cf043c30))

### [1.0.3](https://github.com/xTCry/node-swiftpay-api/compare/v1.0.2...v1.0.3) (2022-01-26)


### Bug Fixes

* **types:** fixed typos `ResponseAccount` ([737ed29](https://github.com/xTCry/node-swiftpay-api/commit/737ed29b5de1fc2746f5f38e9d9713b7f43a5c7f))

### [1.0.2](https://github.com/xTCry/node-swiftpay-api/compare/v1.0.1...v1.0.2) (2022-01-26)


### Features

* **example:** added `payInCreate` to example ([15ef786](https://github.com/xTCry/node-swiftpay-api/commit/15ef78651184ea3352e944f2faa3a48d70430855))


### Bug Fixes

* **types:** fixed `order_id` type ([146f2e0](https://github.com/xTCry/node-swiftpay-api/commit/146f2e0c8a825590a714319d3044b40a792ac900))


### Others

* **api:** removed error logging ([6d20b76](https://github.com/xTCry/node-swiftpay-api/commit/6d20b763f19757b828550dc91ac9c0ebad43f45b))

### [1.0.1](https://github.com/xTCry/node-swiftpay-api/compare/v1.0.0...v1.0.1) (2022-01-26)


### Features

* **api:** added methods `order`, `payInCreate`, `gatePay` ([43b048d](https://github.com/xTCry/node-swiftpay-api/commit/43b048db8619db575b34446dcd1a24806c94648b))


### Others

* **readme:** fixed typos ([5a5753e](https://github.com/xTCry/node-swiftpay-api/commit/5a5753ee90bd5eda77b524e787428a189c68fb2b))

## 1.0.0 (2022-01-26)


### Features

* init repos ([4ef61e6](https://github.com/xTCry/node-swiftpay-api/commit/4ef61e604d52bbdf69fc43e443ede34d6153ac8e))


src/types.ts
/** Статус операции */
export enum ErrorType {
  Unknown = 'Unknown',
  InvalidToken = 'InvalidToken',
  InvalidParams = 'InvalidParams',
}

/** Список доступных валют */
export enum CurrencyType {
  BCH = 'BCH',
  BTC = 'BTC',
  DAI = 'DAI',
  ETH = 'ETH',
  EUR = 'EUR',
  LTC = 'LTC',
  RUB = 'RUB',
  UAH = 'UAH',
  USD = 'USD',
  USDC = 'USDC',
}

export enum SystemType {
  QIWI = 'qiwi',
  CARD = 'card',
  FORM = 'form',
}

export enum SystemIdType {
  XX_30 = 30,
  /** Банковские карты */
  CARD = 40,
  QIWI = 41,
  /** Банковские карты [Trusted] */
  CARD_TRUSTED = 42,
  /** Банковские переводы */
  CARD_TRANSACTION = 43,
  /** Банковские карты [UAH] */
  CARD_UAH = 44,
  /** UDST Tether TRC-20 */
  UDST_TETHER = 45,
}

/** Статус операции */
export enum StatusValue {
  Paid = 'PAID' /* code = 1 */,
  Rejected = 'REJECTED' /* code = 2 */,
  // TODO: add other
}

// Deposits

/** Параметры уведомления о пополнении */
export type ResponsePayment = {
  /**
   * SHA256(order_id + amount + token + shop_id).toUpperCase()
   * Например заказ #123 на 10.43руб мерчанта с ID 22: SHA256(12310.430APIdi7O4mSNzd5ZJiMLEWKw22)
   */
  sign: string;
  // ID в системе SWIFTPAY
  id: number;
  /** ID системы */
  system_id: SystemIdType;
  /** ID пополнения в системе вашего мерчанта */
  order_id: string;
  shop_id: number;
  /** сумма пополнения */
  amount: number;
  amount_in_cur: number;
  cur: number;
  gate_fee: number;

  status: {
    /** Статус пополнения */
    code: number;
    /** Значение статуса */
    value: StatusValue;
  };

  /** дата создания заказа */
  created_at: string;
  /** дата оплаты заказа */
  paid_at: string;
  /** email пользователя */
  email: string;
  /** описание заказа */
  description: string | null;
  /** переданная при создании заказа data */
  data: any;
  /** Комиссия системы Swiftpay за транзакцию */
  commission: number;
  /** Маска карты (Например: 427638******3007) */
  card_mask: string | null;
  phone: string | null;
  token: string;
  hash: string;
};

// Responses

export interface IResponseData<T> {
  // status: 'info' | 'error' | /* ? */ 'success';
  text?: string;
  data: T;
}

export interface IResponseErrorData {
  errCode?: number;
  error: string;
  text?: string;
}

//

export type ResponseAccount = {
  id: number;
  name: string;
  balance: number;
  balanceCurrency: CurrencyType;
  wallets: Record<CurrencyType, number>;
};

export type ResponseStats = Record<'today' | 'yesterday' | 'month' | 'year', { add: number; sub: number }>;

export type ResponseShops = {
  id: number;
  name: string;
  url: string;
  verified: number;
  p2p_balance: number;
}[];

export type ResponseShop = {
  id: number;
  name: string;
  url: string;
  verified: number;
  p2p_balance: number;
  commission: number;
  url_success: string;
  url_redirect: string;
  token: string;
  p2p_notify: null;
  p2p_upincome: number;
  systems: SystemIdType[];
  allSystems: {
    id: SystemIdType;
    name: string;
    commission: number;
  }[];
  balances: null;
};

export type ResponseShopActivate = Boolean;
export type ResponseShopDeactivate = Boolean;
export type ResponseShopDelete = Boolean;

// Orders

type OrderFields = 'id' | 'amount' | 'order_id' | 'created_at' | 'paid_at' | 'shop_id' | 'email' | 'status';
type OrderSortableFields = Exclude<OrderFields, 'email' | 'status'>;
type OrderAllowedQueryFields = OrderFields;
type OrderAllowedOperators = '=' | '>' | '<' | '>=' | '<=' | '!=';

export type ParamsOrders = {
  /** Поле, по которому будет выполнена сортировка */
  sort: OrderSortableFields;
  /** `ASC` - по возрастанию, `DESC` - по убыванию */
  sortType?: 'ASC' | 'DESC';

  /**
   * Если нужно отобразить записи без фильтра, то передайте пустой массив - `data: []`
   *
   * **Пример:**
   * Отобразит только те записи, где дата создания была позже 19.01.2020 и сумма равной 100.
   * @example
   *  [
   *    {
   *      field: 'created_at';
   *      type: '>';
   *      value: '2020-01-19';
   *    },
   *    {
   *      field: 'amount';
   *      type: '=';
   *      value: '100';
   *    },
   *  ]
   * */
  data: {
    field: OrderAllowedQueryFields;
    type: OrderAllowedOperators;
    value: string;
  }[];

  /** Число записей **(от 25 до 250)** */
  limit?: number;
  /**
   * Смещение по записям.
   *
   * При `limit = 25`, `offset = 25` будет второй страницей, `offset = 50` будет третьей и т.д.
   * */
  offset?: number;
};

export type ResponseOrders = {
  id: number;
  shop_id: number;
  ticket: null;
  amount: number;
  status: 1 | 2 | number;
  amount_in_cur: number;
  cur: 1 | number;
  cur_name: CurrencyType;
  created_at: string;
  paid_at: string | null;
  email: string;
  description: string | null;
  log: string;
  token: string;
  order_id: string;
  ticket_status: null;
}[];

export type ResponseOrder = {
  id: number;
  url: string;
  email: string;
  url_redirect: string;
  amount: number;
  to_pay: any | null;
  amount_in_cur: number;
  processed_at: string;
  description: string;
  system_id: SystemIdType;
  status: 0 | 1 | number;
  substatus: 2 | number;
  name: string;
  created_at: string;
  paid_at: string;
  invoice: string | null;
  token: string;
  paid_at_formated: string;
  link: string | null;
  commission: number;
  shop_commission: number;
  income_in_cur: number;
  commission_amount_in_cur: number;
  cur: { name: CurrencyType; precision: number };
  systems: string | null;
  system_name: string | null;
  /** (`true` - **оплачен**, `false` - **неоплачен**) */
  paid: boolean;
  check: string;
};

// Payouts

type PayoutsFields = 'id' | 'amount' | 'system_id' | 'wallet' | 'created_at' | 'paid_at' | 'api' | 'status';
type PayoutsSortableFields = Exclude<PayoutsFields, 'status'>;
type PayoutsAllowedQueryFields = PayoutsFields;
type PayoutsAllowedOperators = '=' | '>' | '<' | '>=' | '<=' | '!=';

export type ParamsPayouts = {
  /** Поле, по которому будет выполнена сортировка */
  sort: PayoutsSortableFields;
  /** `ASC` - по возрастанию, `DESC` - по убыванию */
  sortType?: 'ASC' | 'DESC';

  /**
   * Если нужно отобразить записи без фильтра, то передайте пустой массив - `data: []`
   *
   * **Пример:**
   * Отобразит только те записи, где дата создания была позже 19.01.2020 и сумма равной 100.
   * @example
   *  [
   *    {
   *      field: 'created_at';
   *      type: '>';
   *      value: '2020-01-19';
   *    },
   *    {
   *      field: 'amount';
   *      type: '=';
   *      value: '100';
   *    },
   *  ]
   * */
  data: {
    field: PayoutsAllowedQueryFields;
    type: PayoutsAllowedOperators;
    value: string;
  }[];

  /** Число записей **(от 25 до 250)** */
  limit?: number;
  /**
   * Смещение по записям.
   *
   * При `limit = 25`, `offset = 25` будет второй страницей, `offset = 50` будет третьей и т.д.
   * */
  offset?: number;
};

export type ResponsePayouts = {
  id: number;
  system_id: SystemIdType;
  amount: number;
  amount_in_cur: number;
  cur: 1 | number;
  cur_name: CurrencyType;
  wallet: string;
  created_at: string;
  paid_at: string | null;
  status: 1 | 2 | number;
  api: number;
  usdt: number;
}[];

// Payout

export type ResponsePayout = {};

//

export type ResponsePayoutAccept = {};

export type ResponsePayoutDecline = {};

export type ResponseSystems = {
  id: SystemIdType;
  name: string;
  payout_commission: number;
  payout_fixed_commission: number;
  regexp: string /* RegExp */;
}[];

export type ResponseOrderById = {
  id: number;
  amount: number;
  description: string | null;
  system_id: SystemIdType;
  status: 2 | number;
  created_at: string;
  link: string | null;
  commission: number;
  shop_commission: number;
  /** (`true` - **оплачен**, `false` - **неоплачен**) */
  paid: boolean;
  check: string;
};

export type ResponseCreatePayout = {
  error: 0 | number;
  id: number;
  user_id: number;
  system_id: SystemIdType;
  amount: number;
  wallet: string;
  status: 0 | number;
  created_at: string;
  paid_at: string | null;
  api: 0 | number;
  props_id: null;
};

// CreateOrder

export type ParamsCreateOrder = {
  /** Номер заказа в системе мерчанта. **Должен быть уникальным** */
  order_id: string;
  /** Сумма */
  amount: number;
  /** Описание. *Необязательный параметр* */
  desc?: string;

  /**
   * После успешной оплаты этот объект будет отправлен на ссылку уведомления.
   * *Необязательный параметр*
   * **Макс. количество полей: `5`**
   */
  data?: Record<string, string | number | boolean>;
};

export type ResponseCreateOrder = {
  /** Уникальный номер заказа за базе Swiftpay */
  id: number | string;
  /** Ссылка на оплату */
  link: string;
  info: {
    /** Номер заказа в системе мерчанта. */
    order_id: string;
    /** Сумма */
    amount: string;
    /** Объект, указанный при создании ссылки */
    data: Record<string, string | number | boolean>;
  };
};

// PayInCreate

export type ParamsPayInCreate = {
  /**
   * Номер заказа в системе мерчанта, *заранее созданный и сохраненный в базе обязательный*
   *
   * **Должен быть уникальным**
   * *до 36 симв*
   * */
  order_id: string;
  /** Сумма пополнения. `float`	**от 1 до 15000** */
  amount: number;

  /**
   * Объект, передаваемый на сервер вместе с уведомлением об успешном платеже.
   * **Макс. количество полей: `5`**
   */
  data?: Record<string, string | number | boolean>;

  /** Платежная система */
  system: SystemType;
  /** Email пользователя **от 1 до 255 симв.** */
  email: string;
  /** Ссылка для перенаправления на страницу успеха	**от 1 до 255 симв.** */
  redirect_success: string;
  /** Ссылка для перенаправления на страницу неуспеха	**от 1 до 255 симв.** */
  redirect_fail: string;
};

export type ResponsePayInCreate = {
  /** Уникальный номер заказа за базе Swiftpay */
  id: number;
  /** Ссылка на оплату */
  link: string;
  info: Partial<ParamsPayInCreate>;
};

// GatePay

export type ParamsGatePay = {
  /** Номер заказа в системе мерчанта. **Должен быть уникальным** */
  order_id: string;
  /** Сумма */
  amount: number;

  /**
   * После успешной оплаты этот объект будет отправлен на ссылку уведомления.
   * *Необязательный параметр*
   * **Макс. количество полей: `5`**
   */
  data?: Record<string, string | number | boolean>;

  email: string;
  pan: string;
  expMonth: string;
  expYear: string;
  cvv: string;
  redirect_success: string;
  redirect_fail: string;
  browserData?: {
    browserAcceptHeader: string;
    browserColorDepth: string;
    browserIP: string;
    browserLanguage: string;
    browserScreenHeight: string;
    browserScreenWidth: string;
    browserTZ: string;
    browserUserAgent: string;
    browserJavaEnabled: string;
    windowWidth: string;
    windowHeight: string;
  };
};

export type ResponseGatePay = {
  /** Уникальный номер заказа за базе Swiftpay */
  id: number;
  type: '3DS';
  /** Ссылка на оплату */
  link: string;
};


src/errors.ts
import { ErrorType, IResponseErrorData } from './types';

const getErrorType = ({ error }: IResponseErrorData) => {
  const lowerDesc = error.toLowerCase();
  switch (true) {
    case error === 'Ошибка авторизации':
      return ErrorType.InvalidToken;
    case error === 'Ошибка при валидации параметров':
      return ErrorType.InvalidParams;
    // case lowerDesc.startsWith('Invalid Purse '.toLowerCase()):
    //   return ErrorType.InvalidPurse;
    // TODO: add other errors

    default:
      return ErrorType.Unknown;
  }
};

export class ResponseError extends Error {
  name = this.constructor.name;

  public readonly type: ErrorType;
  public readonly code?: number;

  constructor(readonly body: IResponseErrorData) {
    super();
    this.message = !body.text ? body.error : `${body.error}: ${body.text}`;
    this.code = body.errCode;
    this.type = getErrorType(body);
  }

  get [Symbol.toStringTag](): string {
    return `Error ${this.name}; type=${this.type}${!this.code ? '' : `; code=${this.code}`}`;
  }

  toJSON(): Error & { type: ErrorType; code?: number } {
    return {
      message: this.message,
      name: this.name,
      type: this.type,
      code: this.code,
      stack: this.stack,
    };
  }
}

export class ParamsError extends Error {
  name = this.constructor.name;

  constructor(readonly message: string) {
    super();
    this.message = message;
  }

  get [Symbol.toStringTag](): string {
    return `[Error ${this.name}] ${this.message}`;
  }

  toJSON(): Error {
    return {
      message: this.message,
      name: this.name,
      stack: this.stack,
    };
  }
}


src/api-client.ts
import axios, { AxiosInstance, AxiosRequestConfig, Method } from 'axios';
import * as types from './types';
import { ParamsError, ResponseError } from './errors';
import { sha256 } from './utils';
import { USER_AGENT } from './identity';

export class ApiClient {
  public static readonly API_URL = 'https://api.swiftpay.store';

  private readonly _httpClient: AxiosInstance;

  constructor(public readonly apiKey: string, public readonly shopId?: number, apiOptions: AxiosRequestConfig = {}) {
    this._httpClient = axios.create({
      baseURL: ApiClient.API_URL,
      timeout: 10e3,
      ...apiOptions,
    });
    this._httpClient.defaults.headers.common['User-Agent'] = USER_AGENT;
  }

  public async _callApi<T = unknown>(url: string, method: Method = 'POST', data: any = {}) {
    try {
      const params: any = {};
      if (method == 'GET') {
        params.token = this.apiKey;
      } else {
        data.token = this.apiKey;
      }

      const resp = await this._httpClient.request<types.IResponseData<T>>({ url, params, method, data });
      return resp.data;
    } catch (e: any) {
      if (e.response?.status === 404) {
        throw new ResponseError({ errCode: -1, error: 'Method not found' } as types.IResponseErrorData);
      }
      if (e.response?.data?.error) {
        throw new ResponseError(e.response.data as types.IResponseErrorData);
      }
      throw e;
    }
  }

  /**
   * Проверяет подпись уведомления о выплате
   * @param body Объект уведомления
   * @returns {boolean} Признак валидности
   */
  public checkSignPayment(body: types.ResponsePayment) {
    const compareSign = sha256(`${body.order_id + body.amount + this.apiKey + body.shop_id}`).toUpperCase();

    return body.sign === compareSign;
  }

  /**
   * Получение информации об аккаунте
   */
  public async getAccount() {
    return await this._callApi<types.ResponseAccount>('account', 'POST');
  }

  /**
   * Получение статистики аккаунта
   */
  public async getStats() {
    return await this._callApi<types.ResponseStats>('stats', 'GET');
  }

  /**
   * Получение списка созданных мерчантов
   */
  public async getShops() {
    return await this._callApi<types.ResponseShops>('shops', 'GET');
  }

  /**
   * Получение информации о конкретном мерчанте
   */
  public async getShop(shopId: number = this.shopId!) {
    if (!shopId || Number(shopId) <= 0) {
      throw new ParamsError('shopId should be a number greater than 0');
    }
    return await this._callApi<types.ResponseShop>(`shops/${shopId}`, 'GET');
  }

  /**
   * Включение мерчанта
   */
  public async shopActivate(shopId: number = this.shopId!) {
    if (!shopId || Number(shopId) <= 0) {
      throw new ParamsError('shopId should be a number greater than 0');
    }
    return await this._callApi<types.ResponseShopActivate>(`shop/activate/${shopId}`, 'GET');
  }

  /**
   * Отключение мерчанта
   */
  public async shopDeactivate(shopId: number = this.shopId!) {
    if (!shopId || Number(shopId) <= 0) {
      throw new ParamsError('shopId should be a number greater than 0');
    }
    return await this._callApi<types.ResponseShopDeactivate>(`shop/deactivate/${shopId}`, 'GET');
  }

  /**
   * Удаление мерчанта
   */
  public async shopDelete(shopId: number = this.shopId!) {
    if (!shopId || Number(shopId) <= 0) {
      throw new ParamsError('shopId should be a number greater than 0');
    }
    return await this._callApi<types.ResponseShopDelete>(`shop/delete/${shopId}`, 'GET');
  }

  /**
   * Получение списка пополнений аккаунта с фильтрами
   */
  public async orders(query: types.ParamsOrders) {
    return await this._callApi<types.ResponseOrders>('orders', 'POST', query);
  }

  /**
   * Проверка платежа
   * 
   * [Документация Swiftpay по методу `order`](https://swiftpayru.docs.apiary.io/#reference/4/2//order/:ordertoken)
   */
  public async order(orderToken: string) {
    if (!orderToken || orderToken.length < 10) {
      throw new ParamsError('orderToken required');
    }
    return await this._callApi<types.ResponseOrder>(`order/${orderToken}`, 'GET');
  }

  /**
   * Получение списка выводов аккаунта с фильтрами
   */
  public async payouts(query: types.ParamsPayouts) {
    return await this._callApi<types.ResponsePayouts>('payouts', 'POST', query);
  }

  /**
   * Получение информации о выводе
   */
  public async payout(payoutId: number) {
    if (!payoutId || Number(payoutId) <= 0) {
      throw new ParamsError('payoutId should be a number greater than 0');
    }
    return await this._callApi<types.ResponsePayout>(`payout/${payoutId}`, 'GET');
  }

  /**
   * Подтверждение выплаты (если в настройках аккаунта отключены автовыплаты)
   */
  public async payoutAccept(payoutId: number) {
    if (!payoutId || Number(payoutId) <= 0) {
      throw new ParamsError('payoutId should be a number greater than 0');
    }
    return await this._callApi<types.ResponsePayoutAccept>(`payoutAccept/${payoutId}`, 'GET');
  }

  /**
   * Отклонение оплаты (если в настройках аккаунта отключены автовыплаты)
   */
  public async payoutDecline(payoutId: number) {
    if (!payoutId || Number(payoutId) <= 0) {
      throw new ParamsError('payoutId should be a number greater than 0');
    }
    return await this._callApi<types.ResponsePayoutDecline>(`payoutDecline/${payoutId}`, 'GET');
  }

  /**
   * Получение списка активных систем и их комиссий
   */
  public async systems() {
    return await this._callApi<types.ResponseSystems>('systems', 'GET');
  }

  /**
   * Получение информации о заказе
   */
  public async orderById(orderId: number) {
    if (!orderId || Number(orderId) <= 0) {
      throw new ParamsError('orderId should be a number greater than 0');
    }
    return await this._callApi<types.ResponseOrderById>(`orderById/${orderId}`, 'GET');
  }

  /**
   * Создание ссылки для оплаты
   */
  public async createOrder(params: types.ParamsCreateOrder, shopId: number = this.shopId!) {
    if (!shopId || Number(shopId) <= 0) {
      throw new ParamsError('shopId should be a number greater than 0');
    }
    return await this._callApi<types.ResponseCreateOrder>('createOrder', 'POST', {
      shop_id: shopId,
      ...params,
    });
  }

  /**
   * Создание выплаты
   * @param system_id Id системы
   * @param amount Сумма вывода
   * @param wallet Кошелек для вывода
   */
  public async createPayout(system_id: types.SystemIdType, amount: number, wallet: string) {
    if (!amount || Number(amount) <= 0) {
      throw new ParamsError('amount should be a number greater than 0');
    }
    return await this._callApi<types.ResponseCreatePayout>('createPayout', 'POST', { system_id, amount, wallet });
  }

  /**
   * Создание платежа
   * 
   * [Документация Swiftpay по методу `payInCreate`](https://swiftpayru.docs.apiary.io/#reference/4/0//payin/create)
   */
  public async payInCreate(params: types.ParamsPayInCreate, shopId: number = this.shopId!) {
    if (!shopId || Number(shopId) <= 0) {
      throw new ParamsError('shopId should be a number greater than 0');
    }
    return await this._callApi<types.ResponsePayInCreate>('payIn/create', 'POST', {
      shop_id: shopId,
      ...params,
    });
  }

  /**
   * Создание платежа **по реквизитам карты**
   * 
   * [Документация Swiftpay по методу `gatePay`](https://swiftpayru.docs.apiary.io/#reference/4/1//gate/pay)
   */
  public async gatePay(params: types.ParamsGatePay, shopId: number = this.shopId!) {
    if (!shopId || Number(shopId) <= 0) {
      throw new ParamsError('shopId should be a number greater than 0');
    }
    return await this._callApi<types.ResponseGatePay>('gate/pay', 'POST', {
      shop_id: shopId,
      ...params,
    });
  }
}


example/index.ts
import 'dotenv/config';
import { ApiClient, SystemType } from '../src';

const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));

const main = async () => {
  const api = new ApiClient(
    process.env.WALLET_PROVIDER_SWIFTPAY_API_KEY!,
    Number(process.env.WALLET_PROVIDER_SWIFTPAY_SHOP_ID!),
  );

  let time = Date.now();
  const stats = await api.getStats();
  console.log('stats', stats);
  console.log(`Time: ${Date.now() - time} ms`);

  await sleep(500);

  time = Date.now();
  const account = await api.getAccount();
  console.log('account', account);
  console.log(`Time: ${Date.now() - time} ms`);

  await sleep(500);

  time = Date.now();
  const shops = await api.getShops();
  console.log('shops', shops);
  console.log(`Time: ${Date.now() - time} ms`);

  await sleep(500);

  if (process.env.WALLET_PROVIDER_SWIFTPAY_SHOP_ID) {
    time = Date.now();
    const { data: shop } = await api.getShop();
    console.log('shop', shop);
    console.log(`Time: ${Date.now() - time} ms`);
  }

  await sleep(500);

  // time = Date.now();
  // const { data: createOrder } = await api.createOrder({
  //   amount: 100,
  //   order_id: 4e7 + 2,
  //   data: { userId: 1230, email: 'gmail' },
  // });
  // console.log('createOrder', createOrder);
  // console.log(`Time: ${Date.now() - time} ms`);

  // await sleep(500);

  time = Date.now();
  const orders = await api.orders({ sort: 'id', data: [] });
  console.log('orders', orders);
  console.log(`Time: ${Date.now() - time} ms`);

  await sleep(500);

  time = Date.now();
  const payInCreate = await api.payInCreate({
    order_id: String(4e7 + 3),
    amount: 120,
    system: SystemType.CARD,
    email: `x${123}@test.ru`,
    redirect_success: `https://google.com/?deposit=success`,
    redirect_fail: `https://google.com/?deposit=fail`,
  });
  console.log('payInCreate', payInCreate);
  console.log(`Time: ${Date.now() - time} ms`);
};

main()
  .then()
  .catch((e) => {
    console.error(e);
  });
